src/fsharp/CompileOps.fs

```diff
@@ -4291,22 +4292,29 @@ type TcImports(tcConfigP:TcConfigProvider, initialResolutions:TcAssemblyResoluti
                       with e -> error(Error(FSComp.SR.buildErrorOpeningBinaryFile(filename, e.Message),m))
                 else 
                     tcImports.PrepareToImportReferencedILAssembly (ctok, m, filename, dllinfo)
-            dllinfo,phase2
+            return dllinfo, phase2
+      }
 
-    member tcImports.RegisterAndImportReferencedAssemblies (ctok, nms:AssemblyResolution list) =
+    member tcImports.RegisterAndImportReferencedAssemblies (ctok, nms: AssemblyResolution list) =
+      async {
         CheckDisposed()
 
-        let dllinfos,phase2s = 
-           nms |> List.choose 
-                    (fun nm ->
+        let! r =
+           nms 
+           |> Async.List.choose 
+                (fun nm ->
+                    async {
                         try
-                            Some(tcImports.RegisterAndPrepareToImportReferencedDll (ctok, nm))
+                            let! r = tcImports.RegisterAndPrepareToImportReferencedDll (ctok, nm)
+                            return Some r
                         with e ->
                             errorR(Error(FSComp.SR.buildProblemReadingAssembly(nm.resolvedPath, e.Message),nm.originalReference.Range))
-                            None)
-               |> List.unzip
+                            return None
+                    })
+        let  dllinfos, phase2s = List.unzip r
         let ccuinfos = (List.collect (fun phase2 -> phase2()) phase2s) 
-        dllinfos,ccuinfos
+        return dllinfos, ccuinfos
+      }
       
     member tcImports.DoRegisterAndImportReferencedAssemblies(ctok, nms) = 
```

Comment:
I suspect``DoRegisterAndImportReferencedAssemblies`` should be cancellable, so ``BuildFrameworkTcImports`` and ``BuildNonFrameworkTcImports`` are as well, but ``ImplicitLoadIfAllowed`` should not (i.e. should use RunSynchronously)

BTW one of the problems this approach may have is that the use of ``async`` may interfere with the use of thread statics for ``buildPhase`` and ``errorLogger``, as the execution of the async may reschedule continuations on other threads. 

We could perhaps use a less-rich ``cancellable { ... }`` monad that stays on the same thread, can use thread statics, but does implicit cancellation checks and goes down a cancellation continuation rather than raising an exception.  Or perhaps ``eventually { ... }`` could be adjusted to also carry a cancellation continuation.